plugin-server/src/utils/rate-limiter.ts

```diff
@@ -0,0 +1,51 @@
+import { DateTime } from 'luxon'
+
+import { DB } from './db/db'
+
+export type RateLimitedResource = 'events' | 'recordings'
+
+const CACHE_TTL_SECONDS = 60
+export const RATE_LIMITER_CACHE_KEY = '@posthog-plugin-server/rate-limiter/'
+
+export class RateLimiter {
+    private db: DB
+    private localCache: { [key: string]: string[] } = {}
+    private localCacheExpiresAt: DateTime | null = null
+    private localCacheRefreshingPromise: Promise<void> | null = null
+
+    constructor(db: DB) {
+        this.db = db
+    }
+
+    public async refreshLocalCache(): Promise<void> {
+        if (!this.localCacheExpiresAt || DateTime.utc() > this.localCacheExpiresAt) {
+            if (!this.localCacheRefreshingPromise) {
+                // NOTE: We probably want a timeout here...
+                this.localCacheRefreshingPromise = this.refreshCaches().then(() => {
+                    this.localCacheRefreshingPromise = null
+                })
+            }
+
+            if (!this.localCacheExpiresAt) {
```

Comment:
Comment here might help for the future
```suggestion
            if (!this.localCacheExpiresAt) {
              // We only want to wait for the cache if we haven't at least once loaded the cache
```